README.md
# couchdb_connector

[![Build Status](https://travis-ci.org/locolupo/couchdb_connector.svg)](https://travis-ci.org/locolupo/couchdb_connector)
[![Coverage Status](https://coveralls.io/repos/locolupo/couchdb_connector/badge.svg?branch=master&service=github)](https://coveralls.io/github/locolupo/couchdb_connector?branch=master)
[![Hex.pm](https://img.shields.io/hexpm/v/couchdb_connector.svg?style=flat-square)](https://hex.pm/packages/couchdb_connector)
[![API Docs](https://img.shields.io/badge/api-docs-yellow.svg?style=flat)](https://hexdocs.pm/couchdb_connector/)
[![Deps Status](https://beta.hexfaktor.org/badge/prod/github/locolupo/couchdb_connector.svg)](https://beta.hexfaktor.org/github/locolupo/couchdb_connector)

## Description

A connector for CouchDB, the Erlang-based, JSON document database.

The connector does not implement the protocols defined in Ecto.
Reasons: CouchDB does not support transactions as known in the world of
ACID compliant, relational databases.
The concept of migrations also does not apply to CouchDB.
And since CouchDB does not implement an SQL dialect, the decision was taken
to not follow the standards established by Ecto.

The connector offers create, update and read operations through its
Writer and Reader modules.
Basic support for view operations is provided by the View module.

All create and update operations expect valid JSON documents. All read
operations return JSON strings exactly as they come from CouchDB.

In addition to this representation, Release 0.4 introduced a second format for documents. Users can now also retrieve and ingest documents represented as nested Maps.

The connector also offers functions to manage users and admins. These functions
have been implemented to support testing of authentication and most users will
probably manage users through different tools.

HTTP Basic access authentication (Basic auth) is currently the only supported authentication scheme.

## Supported platforms

The current release of the connector has been tested successfully with Elixir release versions 1.2.6, 1.3.4 and 1.4.0, using Erlang OTP in versions 18.2.1 and 19.2 as well as CouchDB version 1.6.1.

## Installation

The module is [available in Hex](https://hex.pm/packages/couchdb_connector), the package can be installed as follows:

  1. Add couchdb_connector to your list of dependencies in `mix.exs`:

```Elixir
def deps do
  [{:couchdb_connector, "~> 0.5.0"}]
end
```

  2. Ensure couchdb_connector is started before your application:

```Elixir
def application do
  [applications: [:couchdb_connector]]
end
```

## Usage

For the subsequent steps, let's assume that we work in iex and that we define these database properties:

```Elixir
db_props = %{protocol: "http", hostname: "localhost", database: "couchdb_connector_dev", port: 5984}
```

### Authentication

HTTP Basic authentication was first introduced in release version 0.3. Some versions and improvements later, the current best practice is to put the Basic auth credentials into the database properties like so:

```Elixir
db_props = %{protocol: "http", hostname: "localhost", database: "couchdb_connector_dev", port: 5984,
user: "username", password: "secret"}
```

Support for this configuration feature has been introduced in release version 0.4.4. It relieves users from having to add the credentials to each authenticated function call. Authentication can now be dealt with once and in one place only.

### Create a database

```Elixir
Couchdb.Connector.Storage.storage_up(db_props)
```

You should see

```Elixir
{:ok, "{\"ok\":true}\n"}
```

In case the database already exists of course, you would see

```Elixir
{:error,
  "{\"error\":\"file_exists\",\"reason\":\"The database could not be created, the file already exists.\"}\n"}
```

### Write to a database

Now that the database exits, we can create documents in it.

```Elixir
Couchdb.Connector.Writer.create(db_props, "{\"key\": \"value\"}")
```

You should see something similar to this

```Elixir
{:ok,
  "{\"ok\":true,\"id\":\"6b66e9bd59c1c35f3ab51165b10889a7\",\"rev\":\"1-59414e77c768bc202142ac82c2f129de\"}\n",
    [{"Server", "CouchDB/1.6.1 (Erlang OTP/18)"},
    {"Location",
    "http://localhost:5984/couchdb_connector_dev/6b66e9bd59c1c35f3ab51165b10889a7"},
    {"ETag", "\"1-59414e77c768bc202142ac82c2f129de\""},
    {"Date", "Sun, 27 Dec 2015 22:28:15 GMT"},
    {"Content-Type", "text/plain; charset=utf-8"}, {"Content-Length", "95"},
    {"Cache-Control", "must-revalidate"}]}
```

In the previous example, we had CouchDB assign a generated id to the document. This will do for most cases. In case you want to provide an id and you are sure that it does not yet exist in the database, you can do this:

```Elixir
Couchdb.Connector.Writer.create(db_props, "{\"key\": \"value\"}", "unique_id")
```

You should then see something like

```Elixir
{:ok,
  "{\"ok\":true,\"id\":\"unique_id\",\"rev\":\"1-59414e77c768bc202142ac82c2f129de\"}\n",
    [{"Server", "CouchDB/1.6.1 (Erlang OTP/18)"},
    {"Location", "http://localhost:5984/couchdb_connector_dev/unique_id"},
    {"ETag", "\"1-59414e77c768bc202142ac82c2f129de\""},
    {"Date", "Sun, 27 Dec 2015 22:32:45 GMT"},
    {"Content-Type", "text/plain; charset=utf-8"}, {"Content-Length", "72"},
    {"Cache-Control", "must-revalidate"}]}
```

### Write to a database — Input given as Map

Starting with version 0.4, you can now also pass in the document as a Map instead of using the JSON String representation. To do so, make use of the top-level API given in the module Couchdb.Connector:

```Elixir
Couchdb.Connector.create(TestConfig.database_properties, %{"key" => "value"}, "42")
```

The response should look similar to this:

```Elixir
{:ok,
 %{headers: %{"Cache-Control" => "must-revalidate", "Content-Length" => "65",
     "Content-Type" => "text/plain; charset=utf-8",
     "Date" => "Thu, 10 Nov 2016 22:22:18 GMT",
     "ETag" => "\"1-59414e77c768bc202142ac82c2f129de\"",
     "Location" => "http://127.0.0.1:5984/couchdb_connector_dev/42",
     "Server" => "CouchDB/1.6.1 (Erlang OTP/19)"},
   payload: %{"id" => "42", "ok" => true,
     "rev" => "1-59414e77c768bc202142ac82c2f129de"}}}
```

In other words, the connector wraps headers and payload in nested Maps — Cool! Note that the handling is uniform regardless of whether an operation succeeds or fails. An error response will look the same as a success response does, with the exception of the :error atom replacing the :ok atom.

```Elixir
{:error,
  %{headers: => %{...},
    payload: %{"error" => "...", ...}}}
```

### Read from a database

Given we have a document under the id "unique_id" in the database that we created in one of the steps above, the following "GET" should return the desired document.

So let's try

```Elixir
Couchdb.Connector.Reader.get(db_props, "unique_id")
```

You should see something akin to this:

```Elixir
{:ok,
  "{\"_id\":\"unique_id\",\"_rev\":\"1-59414e77c768bc202142ac82c2f129de\",\"key\":\"value\"}\n"}
```

In case you ask for a non existing document, like in this example

```Elixir
Couchdb.Connector.Reader.get(db_props, "wrong_id")
```

You should see something this:

```Elixir
{:error,  "{\"error\":\"not_found\",\"reason\":\"missing\"}\n"}
```

### Read from a database — Response wrapped in a Map

Also starting with version 0.4, you can now retrieve a CouchDB document given as a Map instead of the JSON String representation. To do so, make use of the top-level API given in the module Couchdb.Connector:

```Elixir
Couchdb.Connector.get(TestConfig.database_properties, "42")
```

The response should look similar to this:

```Elixir
{:ok,
 %{"_id" => "42", "_rev" => "1-59414e77c768bc202142ac82c2f129de",
   "key" => "value"}}
```

The error case (document does not exist) looks like this:

```Elixir
{:error, %{"error" => "not_found", "reason" => "missing"}}
```

### Update a document

CouchDB demands that clients pass in the document's current revision to make sure that the update operation occurs on the current version of the document. An update request therefore looks like this:

```Elixir
Couchdb.Connector.Writer.update(db_props, "{\"key\": \"new value\"}", "0a89648ca060...", "1-7b2f4edf07d0...")
```

and the response should be similar to this:

```Elixir
{:ok,
 "{\"ok\":true,\"id\":\"0a89648ca060...\",\"rev\":\"2-7b2f4edf07d0...\"}\n",
 [{"Server", "CouchDB/1.6.1 (Erlang OTP/19)"},
  {"Location",
   "http://127.0.0.1:5984/couchdb_connector_test/0a89648ca060..."},
  {"ETag", "\"2-7b2f4edf07d0...\""},
  {"Date", "Sat, 03 Dec 2016 13:57:07 GMT"},
  {"Content-Type", "text/plain; charset=utf-8"}, {"Content-Length", "95"},
  {"Cache-Control", "must-revalidate"}]}
```

Note that an update increments the revision. In the example above, the new revision now starts with 2, indicating that one update has happened since the document was first created.
Also note that the response does not contain the updated document but only states that the update succeeded.

### Update a document — Response wrapped in a Map

Version 0.4.2 introduced a Map based version of the update API. Let's say we have a document bound to a variable called current. Interacting with the API would then look as follows:

```Elixir
updated = %{current | "key" => "new value"}
{:ok, %{:headers => h, :payload => p}} = Connector.update(db_props, updated)
```

The response would look similar to this:

```Elixir
{:ok,
 %{headers: %{"Cache-Control" => "must-revalidate", "Content-Length" => "95",
     "Content-Type" => "text/plain; charset=utf-8",
     "Date" => "Sat, 03 Dec 2016 14:21:10 GMT",
     "ETag" => "\"2-7b2f4edf07...\"",
     "Location" => "http://127.0.0.1:5984/couchdb_connector_test/8b7b622e37c5c8fa9d6505ecb800197f",
     "Server" => "CouchDB/1.6.1 (Erlang OTP/19)"},
   payload: %{"id" => "8b7b622e37c5...", "ok" => true,
     "rev" => "2-7b2f4edf07..."}}}
```

### Delete a document

In order to delete a document, you have to pass in its current revision, the same way that you saw above for the update calls.

So in response to a call like this:

```Elixir
Couchdb.Connector.Writer.destroy(db_props, "42", "1-9b2e3bcc3752...")
```

You should see a response like this:

```Elixir
"{\"ok\":true,\"id\":\"42\",\"rev\":\"2-9b2e3bcc3752a3...\"}\n"
```

### Delete a document — Response wrapped in a Map

Since release 0.4.2, there is also an implementation of the delete functionality in place that wraps the resopnse in a Map. It is located in the top level module Couchdb.Connector and its API is identical to the JSON/String version:

```Elixir
Couchdb.Connector.destroy(db_props, "42", "1-9b2e3bcc3752...")
```
gives
```
%{"id" => "42", "ok" => true, "rev" => "2-9b2e3bcc3752a3..."}
```

### Create a View

CouchDB [Views](http://guide.couchdb.org/editions/1/en/views.html) are defined in JavaScript and consist of mappers and (optional) reducers. Views are grouped together in CouchDB in what is known as Design Documents.

Let's assume that you want to create one or more Views as part of a seeding process. In order to do so, you can encode your Views in JSON files as follows:

```JSON
{
  "_id" : "_design/example",
  "views" : {
    "by_name" : {
      "map" : "function(doc){ emit(doc.name, doc)}"
    }
  }
}
```

Creating this View can then be done through the connector like this:

```Elixir
{:ok, code} = File.read("path/to/view.json")
{:ok, _} = Couchdb.Connector.View.create_view(db_props, "example", code)
```

where "example" is the name of the design document and code now contains the JavaScript as read from file.

You should see something like

```Elixir
{:ok,
 "{\"ok\":true,\"id\":\"_design/example\",\"rev\":\"1-175ebbcc6e519413aeb640e8fc63424d\"}\n"}
```

### Query a View

Querying a View can be done like this:

```Elixir
{:ok, result} = Couchdb.Connector.View.document_by_key(db_props, "design_name", "view_name", "key")
```

In case the document given by "key" exists, you should see something like
```Elixir
{:ok,
 "{\"total_rows\":3,\"offset\":1,\"rows\":[\r\n{\"id\":\"5c09dbf93fd6226c...\",\"key\":\"key\",..."}
```

otherwise, the response should contain an empty list of rows:
```Elixir
{:ok, "{\"total_rows\":0,\"offset\":0,\"rows\":[\r\n\r\n]}\n"}
```

### Destroy a database

```Elixir
Couchdb.Connector.Storage.storage_down(db_props)
```

You should see

```Elixir
{:ok, "{\"ok\":true}\n"}
```

In case that database never existed, you should see

```Elixir
{:error, "{\"error\":\"not_found\",\"reason\":\"missing\"}\n"}
```

## Next

Love to hear from you. Meanwhile, here are some things I'd like to tackle next:

- pool connections to CouchDB
- enhance view handling and query capabilities
- cookie auth, oauth auth
- attachment support
- improve documentation
- complete function specs


Changelog.md
Version 0.5.0 - released 2017-01-15
-------------

- Remove deprecated functions using individual authentication params.
- Bugfix: Add special handling for integer keys to views to fix #40

Version 0.4.4 - released 2017-01-08
-------------

- Improved authentication handling: You can now add basic auth credentials to db properties, eliminating need to authenticate each function call. Thanks to @leifg
- This change makes a number of authenticated functions redundant. These functions have been deprecated and will be removed in release version 0.5

Version 0.4.3 - released 2016-12-21
-------------

- Add Map based view API and documentation
- Complement some missing specs and test cases
- Fix typo in View related deprecation message
- Add function to create database secured through basic auth. Thanks to @leifg

Version 0.4.2 - released 2016-12-04
-------------

- Add Map based update API and documentation
- Add delegates to Writer module destroy functions in top level Connector module
- Document destroy functions

Version 0.4.1 - released 2016-11-20
-------------

- Bugfix: Add Percent-encoding to query path
- Allow the use of poison 3.0 in addition to 1.5 and 2.0

Version 0.4.0 - released 2016-11-11
-------------

- Provide connector functions that yield and consume documents as Maps
- Support Elixir versions 1.2 and 1.3
- Allow wider range of httpoison versions
- Update development dependencies
- Add and improve documentation

Version 0.3.0 - released 2016-05-13
-------------

- Allow document removal
- Add admin functions for basic user and admin management
- Add basic authentication
- Integrate dialyzer to improve specs and code
- Improve documentation
- Add missing function specs
- Deprecations:
  - Writer.create/2 is now deprecated. Please use Writer.create_generate/2 instead
  - View.document_by_key/5 is now deprecated. Please use one of the new Map-based View.document_by_key functions instead
- Breaking Changes: None in public facing functions

Version 0.2.0 - released 2016-01-04
-------------

First public release


lib/couchdb_connector/admin.ex
defmodule Couchdb.Connector.Admin do
  @moduledoc """
  The Admin module provides functions to create and update users in
  the CouchDB server given by the database properties.

  ## Examples

      db_props = %{protocol: "http", hostname: "localhost",database: "couchdb_connector_test", port: 5984}
      %{database: "couchdb_connector_test", hostname: "localhost", port: 5984, protocol: "http"}

      Couchdb.Connector.Admin.create_user(db_props, "jan", "relax", ["couchdb contributor"])
      {:ok,
       "{\\"ok\\":true,\\"id\\":\\"org.couchdb.user:jan\\",\\"rev\\":\\"1-1d509578d1bc8cba3a6690fca5e7a9fd\\"}\\n",
       [{"Server", "CouchDB/1.6.1 (Erlang OTP/18)"},
        {"Location", "http://localhost:5984/_users/org.couchdb.user:jan"},
        {"ETag", "\\"1-1d509578d1bc8cba3a6690fca5e7a9fd\\""},
        {"Date", "Thu, 31 Mar 2016 21:50:04 GMT"},
        {"Content-Type", "text/plain; charset=utf-8"}, {"Content-Length", "83"},
        {"Cache-Control", "must-revalidate"}]}

      Couchdb.Connector.Admin.create_user(db_props, "jan", "relax", ["couchdb contributor"])
      {:error,
        "{\\"error\\":\\"conflict\\",\\"reason\\":\\"Document update conflict.\\"}\\n",
        [{"Server", "CouchDB/1.6.1 (Erlang OTP/18)"},
         {"Date", "Thu, 31 Mar 2016 21:50:06 GMT"},
         {"Content-Type", "text/plain; charset=utf-8"}, {"Content-Length", "58"},
         {"Cache-Control", "must-revalidate"}]}

      Couchdb.Connector.Admin.user_info(db_props, "jan")
      {:ok,
        "{\\"_id\\":\\"org.couchdb.user:jan\\",\\"_rev\\":\\"1-...\\",
          \\"password_scheme\\":\\"pbkdf2\\",\\"iterations\\":10,\\"type\\":\\"user\\",
          \\"roles\\":[\\"couchdb contributor\\"],\\"name\\":\\"jan\\",
          \\"derived_key\\":\\"a294518...\\",\\"salt\\":\\"70869...\\"}\\n"}

      Couchdb.Connector.Admin.destroy_user(db_props, "jan")
      {:ok,
        "{\\"ok\\":true,\\"id\\":\\"org.couchdb.user:jan\\",\\"rev\\":\\"2-429e8839208ed64cd58eae75957cc0d4\\"}\\n"}

      Couchdb.Connector.Admin.user_info(db_props, "jan")
      {:error, "{\\"error\\":\\"not_found\\",\\"reason\\":\\"deleted\\"}\\n"}

      Couchdb.Connector.Admin.destroy_user(db_props, "jan")
      {:error, "{\\"error\\":\\"not_found\\",\\"reason\\":\\"deleted\\"}\\n"}

  """

  alias Couchdb.Connector.Types
  alias Couchdb.Connector.Headers
  alias Couchdb.Connector.UrlHelper
  alias Couchdb.Connector.ResponseHandler, as: Handler

  @doc """
  Create a new user with given username, password and roles. In case of success,
  the function will respond with {:ok, body, headers}. In case of failures (e.g.
  if user already exists), the response will be {:error, body, headers}.
  Please note that the credentials used in db_properties must be the credentials
  of an admin user.
  """
  @spec create_user(Types.db_properties, Types.user_info, Types.user_roles)
    :: {:ok, String.t, Types.headers} | {:error, String.t, Types.headers}
  def create_user(db_props, user_auth, roles) do
    db_props
    |> UrlHelper.user_url(user_auth[:user])
    |> do_create_user(user_to_json(user_auth, roles))
    |> Handler.handle_put(:include_headers)
  end

  defp do_create_user(url, json) do
    HTTPoison.put! url, json, [Headers.json_header]
  end

  defp user_to_json(user_auth, roles) do
    Poison.encode! %{"name" => user_auth[:user],
                     "password" => user_auth[:password],
                     "roles" => roles,
                     "type" => "user"}
  end

  @doc """
  Create a new admin with given username and password. In case of success,
  the function will respond with an empty body. In case of failures (e.g.
  if admin already exists), the response will be {:error, body, headers}.
  """
  @spec create_admin(Types.db_properties, Types.user_info)
    :: {:ok, String.t, Types.headers} | {:error, String.t, Types.headers}
  def create_admin(db_props, admin_info) do
    db_props
    |> UrlHelper.admin_url(admin_info[:user])
    |> do_create_admin(admin_info[:password])
    |> Handler.handle_put(:include_headers)
  end

  defp do_create_admin(url, password) do
    HTTPoison.put! url, "\"" <> password <> "\"", [Headers.www_form_header]
  end

  @doc """
  Returns the public information for the given user or an error in case the
  user does not exist.
  """
  @spec user_info(Types.db_properties, String.t)
    :: {:ok, String.t} | {:error, String.t}
  def user_info(db_props, username) do
    db_props
    |> UrlHelper.user_url(username)
    |> HTTPoison.get!
    |> Handler.handle_get
  end

  @doc """
  Returns hashed information for the given admin or an error in case the admin
  does not exist or if the given credentials are wrong.
  """
  @spec admin_info(Types.db_properties)
    :: {:ok, String.t} | {:error, String.t}
  def admin_info db_props do
    db_props
    |> UrlHelper.admin_url(db_props[:user])
    |> HTTPoison.get!
    |> Handler.handle_get
  end

  @doc """
  Deletes the given user from the database server or returns an error in case
  the user cannot be found. Requires admin basic auth credentials.
  """
  @spec destroy_user(Types.db_properties, String.t)
  :: {:ok, String.t} | {:error, String.t}
  def destroy_user(db_props, username) do
    case user_info(db_props, username) do
      {:ok, user_json} ->
        user = Poison.decode! user_json
        do_destroy_user(db_props, username, user["_rev"])
      error -> error
    end
  end

  defp do_destroy_user(db_props, username, rev) do
    db_props
    |> UrlHelper.user_url(username)
    |> do_http_delete(rev)
    |> Handler.handle_delete
  end

  defp do_http_delete(url) do
    HTTPoison.delete! url
  end

  defp do_http_delete(url, rev) do
    HTTPoison.delete! url <> "?rev=#{rev}"
  end

  @doc """
  Deletes the given admin from the database server or returns an error in case
  the admin cannot be found.
  """
  @spec destroy_admin(Types.db_properties, String.t, String.t)
    :: {:ok, String.t} | {:error, String.t}
  def destroy_admin(db_props, username, password) do
    db_props
    |> UrlHelper.admin_url(username, password)
    |> do_http_delete
    |> Handler.handle_delete
  end


  @doc """
  Set the security object for a given database. Security object includes admins
  and members for the database.
  """
  # TODO: add user roles
  @spec set_security(Types.db_properties, list(String.t), list(String.t))
    :: {:ok, String.t} | {:error, String.t}
  def set_security(db_props, admins, members) do
    db_props
    |> UrlHelper.security_url
    |> do_set_security(security_to_json(admins, members))
    |> Handler.handle_put
  end

  defp do_set_security(url, json) do
    HTTPoison.put! url, json, [Headers.json_header]
  end

  defp security_to_json(admins, members) do
    Poison.encode!(
    %{:admins  => %{:names => admins,  :roles => ["admins"]},
      :members => %{:names => members, :roles => ["members"]}
    })
  end
end
